src/gluonts/model/renewal/_predictor.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import itertools
+from functools import partial
+from pathlib import Path
+from typing import List, Optional, Callable, Iterator
+
+import mxnet as mx
+import numpy as np
+
+from gluonts.core.component import DType
+from gluonts.core.serde import load_json
+from gluonts.dataset.common import DataEntry, Dataset
+from gluonts.dataset.loader import InferenceDataLoader
+from gluonts.model.forecast import Forecast
+from gluonts.model.forecast_generator import (
+    ForecastGenerator,
+    SampleForecastGenerator,
+)
+from gluonts.mx.batchify import batchify
+from gluonts.mx.context import get_mxnet_context
+from gluonts.mx.model.predictor import RepresentableBlockPredictor
+from gluonts.transform import Transformation
+
+
+class DeepRenewalProcessSampleOutputTransform:
+    """
+    Convert a deep renewal process sample that is composed of dense interval-size
+    samples to a sparse forecast. In practice, this often only means taking the
+    first few time steps of each sample trajectory and converting them to the
+    sparse (intermittent) representation. Converts a (N, S, 2, T) array corresponding
+    to the interval-size format to a (N, S, T) array.
+    """
+
+    def __call__(self, entry: DataEntry, output: np.ndarray) -> np.ndarray:
+        ia_times, sizes = output[:, :, 0], output[:, :, 1]
+        batch_size, num_samples, max_time = ia_times.shape
+        max_time = ia_times.shape[-1]
+        out = np.zeros_like(ia_times)
+
+        times = np.cumsum(ia_times, axis=-1) - 1
+
+        for i, j in itertools.product(range(batch_size), range(num_samples)):
+            t, s = times[i, j], sizes[i, j]
+            ix = t[t < max_time].astype(int)
+            out[i, j, ix] = s[: len(ix)]
+
+        return out
+
+
+class DeepRenewalProcessPredictor(RepresentableBlockPredictor):
```

Comment:
It's the the data loader. The Predictor objects fix the data loader and don't give a way to tamper with it from the outside. This I think is the least intrusive way to make it work with a data loader that is aware of padding.

Reply:
I see. If I read through the class hierarchy correctly, maybe we can expose an option in `GluonEstimator` to specify the stacking function? It could default to `None`, in which case it is set to

```python
partial(batchify, ctx=self.ctx, dtype=self.dtype)
```

and any custom estimator type can set it to something else. In your case it would be

```python
partial(batchify, ctx=self.ctx, dtype=self.dtype, variable_length=True, is_right_pad=False)
```

Reply:
There's actually another small minor thing. This class re-exposes `input_names`. 

In general, this is because the DeepRenewalNetwork contains parameters following the HybridBlock convention. When this is the case, these parameters have to appear in the `hybrid_forward` signatures, which breaks `get_hybrid_forward_input_names`. 

If re-exposing this is also OK, I can change `GluonPredictor` (which is what I assume you meant).

Reply:
Yeah I think that makes sense. It seems like `RepresentablePredictor` would require the change there. For backward compat, we can again put that to `None` in which the class falls back to using `get_hybrid_forward_input_names`.

Sorry about the inconvenient there, but I guess it's one of those situations in which abstraction can improve only be throwing more and more edge cases at them.

Reply:
No inconvenience at all. I started implementing this change, however I think I'm better off sticking to this version for reasons outlined below. 

This class has two custom things that could be passed as constructor arguments to the super class: the output transform and the stack_fn. However, if these are passed, they won't be serializable in the GluonTS manner. In fact, I think OutputTransform is anyway ignored by the `GluonPredictor` (this would be pretty bad for this predictor); and stack_fn with `partial` will also not be JSON serializable. For consistent serde, I think we could stick to this version. What do you say?

The doc build fail is something general I think, both earlier versions of this PR (that passed) and other PRs are affected.

Reply:
> For consistent serde, I think we could stick to this version. What do you say?

Yes, maybe we do that for now, but open an issue pointing to this? It would be good to sort this out in general, and remove a lot of duplication (not only here I think)

Reply:
A short issue can be found here: https://github.com/awslabs/gluon-ts/issues/1469